non-fungible-token/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+        }
+        if spender == owner {
+            panic!("NonFungibleToken: Approval to current owner");
+        }
+        if !self.is_token_owner(token_id, owner) {
+            panic!("NonFungibleToken: is not owner");
+        }
+
+        self.token_approvals.insert(token_id, *spender);
+
+        let approve_token = Approve {
+            owner: H256::from_slice(owner.as_ref()),
+            spender: H256::from_slice(spender.as_ref()),
+            token_id,
+        };
+
+        msg::reply(
+            Event::Approval(approve_token),
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+
+    fn approve_for_all(&mut self, owner: &ActorId, operator: &ActorId, approved: bool) {
+        if operator == &ZERO_ID {
+            panic!("NonFungibleToken: Approval for a zero address");
+        }
+
+        self.operator_approval
+            .entry(*owner)
+            .or_default()
+            .insert(*operator, approved);
+
+        let approve_operator = ApproveForAll {
+            owner: H256::from_slice(owner.as_ref()),
+            operator: H256::from_slice(operator.as_ref()),
+            approved,
+        };
+
+        msg::reply(
+            Event::ApprovalForAll(approve_operator),
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+
+    fn balance_of(&self, account: &ActorId) {
+        let balance = *self.balances.get(account).unwrap_or(&U256::zero());
+        msg::reply(
+            Event::BalanceOf(balance),
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+
+    fn owner_of(&self, token_id: U256) {
+        let owner = self.owner_by_id.get(&token_id).unwrap_or(&ZERO_ID);
+        msg::reply(
+            Event::OwnerOf(H256::from_slice(owner.as_ref())),
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+}
+
+impl NonFungibleToken {
+    pub fn is_token_owner(&self, token_id: U256, account: &ActorId) -> bool {
+        account == self.owner_by_id.get(&token_id).unwrap_or(&ZERO_ID)
+    }
+
+    pub fn authorized_actor(&self, token_id: U256, account: &ActorId) -> AuthAccount {
+        let owner = self.owner_by_id.get(&token_id).unwrap_or(&ZERO_ID);
+        if owner == account {
+            return AuthAccount::Owner;
+        }
+        if self.token_approvals.get(&token_id).unwrap_or(&ZERO_ID) == account {
+            return AuthAccount::ApprovedActor;
+        }
+        if *self
+            .operator_approval
+            .get(owner)
+            .unwrap_or(&BTreeMap::<ActorId, bool>::default())
+            .get(account)
+            .unwrap_or(&false)
+        {
+            return AuthAccount::Operator;
+        }
+        AuthAccount::None
+    }
+
+    pub fn exists(&self, token_id: U256) -> bool {
+        self.owner_by_id.contains_key(&token_id)
+    }
+}
+
+#[derive(Debug, Encode, Decode, TypeInfo)]
+pub struct Approve {
+    owner: H256,
```

Comment:
Suggestion: AFAIK `owner : ActorId` works instead of converting from H256. Can be done later if deemed profitable.

Reply:
Exactly, I will change that